godot-core/src/builtin/quaternion.rs

```diff
@@ -130,70 +138,115 @@ impl Quaternion {
-        }
+        let interpolated = self.as_inner().slerp(to, weight.as_f64());
 
-        if 1.0 - cosom > real::CMP_EPSILON {
-            omega = cosom.acos();
-            sinom = omega.sin();
-            scale0 = ((1.0 - weight) * omega).sin() / sinom;
-            scale1 = (weight * omega).sin() / sinom;
-        } else {
-            scale0 = 1.0 - weight;
-            scale1 = weight;
-        }
+        // Godot returns default if you give it quaternions that are not normalized. This means we can check for default
+        // Then check if we should panic.
+        let normalized_inputs = !interpolated.is_default()
+            || interpolated.is_default() && self.ensure_normalized(&[&to]);
+        assert!(normalized_inputs, "Slerp requires normalized quaternions");
 
-        scale0 * self + scale1 * to1
+        interpolated
     }
 
+    /// # Panics
+    /// If either quaternion is not normalized.
     pub fn slerpni(self, to: Self, weight: real) -> Self {
-        let dot = self.dot(to);
-        if dot.abs() > 0.9999 {
-            return self;
-        }
-        let theta = dot.acos();
-        let sin_t = 1.0 / theta.sin();
-        let new_factor = (weight * theta).sin() * sin_t;
-        let inv_factor = ((1.0 - weight) * theta).sin() * sin_t;
-
-        inv_factor * self + new_factor * to
-    }
-
-    // pub fn spherical_cubic_interpolate(self, b: Self, pre_a: Self, post_b: Self, weight: real) -> Self {}
-    // TODO: Implement godot's function in Rust
-    /*
-        pub fn spherical_cubic_interpolate_in_time(
-            self,
-            b: Self,
-            pre_a: Self,
-            post_b: Self,
-            weight: real,
-            b_t: real,
-            pre_a_t: real,
-            post_b_t: real,
-        ) -> Self {
-        }
-    */
+        let interpolated = self.as_inner().slerpni(to, weight.as_f64());
+
+        // Godot returns default if you give it quaternions that are not normalized. This means we can check for default
+        // Then check if we should panic.
+        let normalized_inputs = !interpolated.is_default()
+            || interpolated.is_default() && self.ensure_normalized(&[&to]);
+        assert!(normalized_inputs, "Slerpni requires normalized quaternions");
+
+        interpolated
+    }
+
+    /// # Panics
+    /// If any quaternions are not normalized.
+    pub fn spherical_cubic_interpolate(
+        self,
+        b: Self,
+        pre_a: Self,
+        post_b: Self,
+        weight: real,
+    ) -> Self {
+        let interpolated =
+            self.as_inner()
+                .spherical_cubic_interpolate(b, pre_a, post_b, weight.as_f64());
+
+        // Godot returns default if you give it quaternions that are not normalized. This means we can check for default
+        // Then check if we should panic.
+        let normalized_inputs = !interpolated.is_default()
+            || interpolated.is_default() && self.ensure_normalized(&[&b, &pre_a, &post_b]);
```

Comment:
```suggestion
        let normalized_inputs = !interpolated.is_default()
            || self.ensure_normalized(&[&b, &pre_a, &post_b]);
```
Logically, `interpolated.is_default()` is implied if `!interpolated.is_default()` is false and the 2nd operand of `||` is evaluated.

Reply:
Oh yeah 🤦‍♂️